### PR TITLE
[MM-18639] Fix search highlighting for phrases

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

Fixes search highlighting for phrases (double quoted strings).

#### Ticket

https://mattermost.atlassian.net/browse/MM-18639

### CODE REVIEW

File: utils/text_formatting.jsx [hunk 1]
```diff
@@ -49,13 +49,15 @@ export function formatText(text, inputOptions) {
...
+    const hasPhrases = (/"([^"]*)"/).test(options.searchTerm);
 
-    if (options.searchMatches) {
+    if (options.searchMatches && !hasPhrases) {
         options.searchPatterns = options.searchMatches.map(convertSearchTermToRegex);
     } else {
-        options.searchPatterns = parseSearchTerms(options.searchTerm).map(convertSearchTermToRegex);
+        options.searchPatterns = parseSearchTerms(options.searchTerm).map(convertSearchTermToRegex).sort((a, b) => {
+            return b.term.length - a.term.length;
+        });
```
Comment (streamer45) [inline utils/text_formatting.jsx]: Here I added a sort by length to make sure the order of input search terms won't affect the highlighted result.

File: utils/text_formatting.jsx [hunk 2]
```diff
@@ -49,13 +49,15 @@ export function formatText(text, inputOptions) {
     }
 
     let output = text;
-
     const options = Object.assign({}, inputOptions);
+    const hasPhrases = (/"([^"]*)"/).test(options.searchTerm);
```
Comment (crspeller) [inline utils/text_formatting.jsx]: Should we be using the xregexp like the others here?

Comment (streamer45) [reply utils/text_formatting.jsx]: That's a good question. I am not familiar with XRegExp. That expression is practically the same that appears in `parseSearchTerms` which I re-used.

https://github.com/mattermost/mattermost-webapp/blob/cbf67e07f5390540d0cb74b51a7ce53cd61ccd14/utils/text_formatting.jsx#L410

Comment (hmhealey) [reply utils/text_formatting.jsx]: XRegExp is only needed if you're using regex features that aren't supported natively in JS. This is fine using a regular regular expression

File: utils/text_formatting.jsx [hunk 3]
```diff
@@ -49,13 +49,15 @@ export function formatText(text, inputOptions) {
...
 
-    if (options.searchMatches) {
+    if (options.searchMatches && !hasPhrases) {
```
Comment (hmhealey) [inline utils/text_formatting.jsx]: I don't think it's too much of a problem, but this is going to break fuzzy highlighting when using phrases. For example, `"row boat"` will match `row boat` and `row boating`, but this change will cause it to only highlight `row boat`

Comment (streamer45) [reply utils/text_formatting.jsx]: That's interesting. When I search `"row boat"` I assume it to only match that exact phrase and not a superset of it. Not sure why `ElasticSearch` behaves like that or if this is expected. I'll investigate further.

**Pull request merged**